Update GA4 `GET:report` to allow sorting by dimensions.

techanvil opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #6513 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

techanvil commented on includes/Modules/Analytics_4.php:
```diff
@@ -1173,39 +1174,36 @@ function ( $metric_def ) {
-
-				if ( empty( $order_def['fieldName'] ) || empty( $order_def['sortOrder'] ) ) {
+				$order_by = new Google_Service_AnalyticsData_OrderBy();
+				$order_by->setDesc( isset( $order_def['desc'] ) ? $order_def['desc'] : false );
```
Ensure the value is a boolean:

```suggestion
				$order_by->setDesc( ! empty( $order_def['desc'] ) );
```

hussain-t reviewed (changes requested):
Thanks, @techanvil. Nice one. The changes are looking good. However, the `assets/js/modules/analytics-4/utils/data-mock.js` hasn’t been modified to accommodate the changes. Can you check that?

techanvil commented:
> Thanks, @techanvil. Nice one. The changes are looking good. However, the `assets/js/modules/analytics-4/utils/data-mock.js` hasn’t been modified to accommodate the changes. Can you check that?

Thanks for spotting that, @hussain-t! I was wondering about how much quicker it was to implement than the estimate, that explains it.

I've updated the JS mock data utility with those changes.

Additionally, while doing so I noticed a subtle error in the way it generates date ranges, so I fixed that here as well (I figured it would be best to fix now so we can unblock #6216 and the following GA4 reporting issues sooner rather than creating another dependency issue to fix it). The fix can be seen in commit 631908e131ed4a15808b1d8a6644adbd378c267a, with the following commit 5217c74d9fa2f6fc77999356fe01f20e2c26091c to simplify things a bit.

hussain-t commented:
>Additionally, while doing so I noticed a subtle error in the way it generates date ranges, so I fixed that here as well (I figured it would be best to fix now so we can unblock https://github.com/google/site-kit-wp/issues/6216 and the following GA4 reporting issues sooner rather than creating another dependency issue to fix it). The fix can be seen in commit https://github.com/google/site-kit-wp/commit/631908e131ed4a15808b1d8a6644adbd378c267a, with the following commit https://github.com/google/site-kit-wp/commit/5217c74d9fa2f6fc77999356fe01f20e2c26091c to simplify things a bit.

Thanks for spotting and fixing this, @techanvil 👍

eugene-manuilov commented on includes/Modules/Analytics_4.php:
```diff
@@ -1173,39 +1174,36 @@ function ( $metric_def ) {
 	 * @param array|null $orderby Data request orderby value.
 	 * @return Google_Service_AnalyticsData_OrderBy[] An array of AnalyticsData OrderBy objects.
 	 */
 	protected function parse_reporting_orderby( $orderby ) {
-		if ( empty( $orderby ) || ! is_array( $orderby ) ) {
+		if ( empty( $orderby ) || ! is_array( $orderby ) || ! wp_is_numeric_array( $orderby ) ) {
 			return array();
 		}
```
It is still possible to use just a single object as orderBy for a report in javascript. With this change, it will be ignored. We should not return an empty array if the `$orderby` variable is a single order object, we should convert it to an array of objects:

```php
if ( empty( $orderby ) || ! is_array( $orderby ) ) {
    return array();
}

if ( ! wp_is_numeric_array( $orderby ) ) {
    $orderby = array( $orderby );
}
```

eugene-manuilov commented on includes/Modules/Analytics_4.php:
```diff
@@ -1173,39 +1174,36 @@ function ( $metric_def ) {
-
-				if ( empty( $order_def['fieldName'] ) || empty( $order_def['sortOrder'] ) ) {
+				$order_by = new Google_Service_AnalyticsData_OrderBy();
+				$order_by->setDesc( ! empty( $order_def['desc'] ) );
```
There is no `desc` property in the order object, we currently still use `sortOrder` with either `ASCENDING` or `DESCENDING` value. Please, make it compatible with the current data structure.

techanvil replied:
Please see my reply to your comment above.

eugene-manuilov commented on includes/Modules/Analytics_4.php:
```diff
@@ -1173,39 +1174,36 @@ function ( $metric_def ) {
+				$order_by = new Google_Service_AnalyticsData_OrderBy();
+				$order_by->setDesc( ! empty( $order_def['desc'] ) );
+
+				if ( isset( $order_def['metric'] ) && isset( $order_def['metric']['metricName'] ) ) {
+					$metric_order_by = new Google_Service_AnalyticsData_MetricOrderBy();
+					$metric_order_by->setMetricName( $order_def['metric']['metricName'] );
+					$order_by->setMetric( $metric_order_by );
+				} elseif ( isset( $order_def['dimension'] ) && isset( $order_def['dimension']['dimensionName'] ) ) {
+					$dimension_order_by = new Google_Service_AnalyticsData_DimensionOrderBy();
+					$dimension_order_by->setDimensionName( $order_def['dimension']['dimensionName'] );
+					$order_by->setDimension( $dimension_order_by );
+				} else {
```
Currently we don't support neither `metric` nor `dimension` properties in the order by object. The only available property with a field name is `fieldName`. I think we can consider it to be a dimension name because as far as I remember we sort everything using the `date` dimension only (i might be wrong, don't remember exactly). Let's make it compatible with the current approach and leave a comment and maybe create a separate ticket to allow `metric`/`dimension` properties in the sort object for Analytics-4 reports.

eugene-manuilov reviewed (changes requested):
Thanks, @techanvil. Left a few comments for you. Please, take a look and let me know if I am missing something.

eugene-manuilov commented on assets/js/util/report-validation.js:
```diff
@@ -99,6 +104,52 @@ export function isValidOrders( orders ) {
+		}
+
+		if ( order.dimension ) {
+			if ( order.metric ) {
+				return false;
+			}
```
This check is dead code because the `metric` property is already processed in the previous condition. We can remove it.

techanvil replied:
Good point, thanks! Removed :+1:

eugene-manuilov commented on assets/js/util/report-validation.js:
```diff
@@ -99,6 +104,52 @@ export function isValidOrders( orders ) {
+		}
+
+		if ( order.metric ) {
+			if ( order.dimension ) {
+				return false;
+			}
+			return typeof order.metric?.metricName === 'string';
```
How about this:
```suggestion
			return ! order.dimension && typeof order.metric?.metricName === 'string';
```

techanvil replied:
SGTM, have updated.

techanvil commented on assets/js/util/report-validation.js:
```diff
@@ -99,6 +104,52 @@ export function isValidOrders( orders ) {
+ * @param {Object[]} orders The order definitions to check.
+ * @return {boolean} TRUE if order definitions are valid, otherwise FALSE.
+ */
+export function isValidOrdersGA4( orders ) {
```
Thanks for pointing that out, I hadn't spotted the GA4-specific report utils. Have moved :+1:

eugene-manuilov commented on assets/js/modules/analytics-4/utils/report-validation.test.js:
```diff
@@ -157,4 +158,135 @@ describe( 'Analytics 4 Reporting API validation', () => {
 	} );
+
+	describe( 'isValidOrders', () => {
+		it( 'should return TRUE if an array of valid order objects is passed', () => {
```
All these tests repeat the same logic. Let's group them into one table test using `it.each( ... )( ... )` approach.

techanvil replied:
Sounds good :+1:

eugene-manuilov commented on assets/js/modules/analytics-4/utils/report-validation.test.js:
```diff
@@ -157,4 +158,135 @@ describe( 'Analytics 4 Reporting API validation', () => {
+						},
+					},
+				] )
+			).toBeTruthy();
```
We need to use `.toBe( true )` because `.toBeTruthy()` doesn't fail the test if non-empty value is passed. So, I think  it would be better to be more precise.

techanvil replied:
Good point. I had used `toBeTruthy()` as that's what the existing UA `isValidOrders()` tests use and I wanted to be consistent. I didn't realise that in fact the UA `isValidOrders()` doesn't always return a boolean and can in fact return  the result of `String.prototype.match()`:

https://github.com/google/site-kit-wp/blob/5ef7abacd409518ad9dfe49f553e9a841e20383e/assets/js/util/report-validation.js#L82-L85

I've created an issue to update UA `isValidOrders()` to always return a boolean, although it's pretty low priority. https://github.com/google/site-kit-wp/issues/6553

eugene-manuilov commented on assets/js/modules/analytics-4/utils/report-validation.test.js:
```diff
@@ -157,4 +158,135 @@ describe( 'Analytics 4 Reporting API validation', () => {
+		it( 'should return FALSE if an array is not passed', () => {
+			expect( isValidOrders( null ) ).toBeFalsy();
+			expect( isValidOrders( 'test' ) ).toBeFalsy();
+			expect( isValidOrders( { test: 123 } ) ).toBeFalsy();
```
Same here, we need to use `.toBe( false )` instead of `.toBeFalsy()`.

techanvil replied:
Thanks, and as above.

eugene-manuilov reviewed (changes requested):
Thanks, @techanvil. The last comments from me. Once they are addressed, we are good to go.

eugene-manuilov reviewed (approved):
Thanks, @techanvil. Looks solid to me! Great work!
